### ISSUE TITLE
Our cache-busting logic for our `app.bundle.js` is faulty

### ISSUE THREAD

**Issue by ggilmore** [sourcegraph/sourcegraph#19471]
# tl;dr

tl;dr - It's possible for Cloudflare to cache the "wrong" (old) version of `app.bundle.js` if the user makes a request at the same time that a deployment rollout is occurring. We should give `app.bundle.js` a unique filename based off of its contents (ex: `app.bundle-abcd1234.js` ) to fix this. 

---

Our cache-busting logic for our `app.bundle.js` is faulty. 

# Background - app.bundle.js and chunks

`app.bundle.js` is the main entry-point to the Sourcegraph web app. `app.bundle.js` dynamically loads in various chunks - depending on what action the user is taking in the web app. These chunks are given names that are derived from of a hash of their contents:

[https://sourcegraph.com/github.com/sourcegraph/sourcegraph@71cb97d88538551e961397a93d74321a8d103d0f/-/blob/client/web/webpack.config.js#L81:35](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@71cb97d88538551e961397a93d74321a8d103d0f/-/blob/client/web/webpack.config.js#L81:35)

```jsx
output: {
    ...
    chunkFilename: 'scripts/[id]-[contenthash].chunk.js',
    ...
  },
```

As an example, if you visit a new page like the "User settings" page,  `app.bundle.js` can make a request to load `.assets/scripts/71-cdb678025685840a93a2.chunk.js` .

 

`app.bundle.js` contains a listing of what chunk names are available: 

```jsx
(l.e = function (e) {
    var t = [],
      n = r[e];
    if (0 !== n)
      if (n) t.push(n[2]);
      else {
        var o = new Promise(function (t, l) {
          n = r[e] = [t, l];
        });
        t.push((n[2] = o));
        var i,
          a = document.createElement("script");
        (a.charset = "utf-8"),
          (a.timeout = 120),
          l.nc && a.setAttribute("nonce", l.nc),
          (a.src = (function (e) {
            return (
              l.p +
              "scripts/" +
              e +
              "-" +
              {
                0: "1397e3389b780aa0da77",
                1: "036c00c084daace320d0",
                2: "42f362f326f0b5afa6cc",
                3: "6c3fae8748e172264eeb",
                4: "b501b34fc7966b0495de",
                5: "3287cf21b9009430cef1",
                6: "d0e1958255e663e5306a",
                7: "50e760302987c0132183",
                8: "a02e01de5a9a2a299205",
                9: "86fc4cdf9250e756f55a",
                10: "ecf2aff1d933e65ec2ad",
```

Whenever the web app contents are updated:

1. The chunk files are given new names to reflect the changes in their contents
2. The newly generated version of `app.bundle.js` gets an updated listing that contains these new chunk names 

# Caching

So, how does Cloudflare handle caching these assets? 

First, here is the [relevant code path](https://github.com/sourcegraph/sourcegraph/blob/71cb97d88538551e961397a93d74321a8d103d0f/cmd/frontend/internal/app/assetsutil/handler.go#L50-L64) in `frontend` that handles serving these assets: 
```go
func Mount(mux *http.ServeMux) {
	const urlPathPrefix = "/.assets"

	fs := httpgzip.FileServer(assets.Assets, httpgzip.FileServerOptions{DisableDirListing: true})
	mux.Handle(urlPathPrefix+"/", http.StripPrefix(urlPathPrefix, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		// ... (snipped for brevity)
		
		// Only cache if the file is found. This avoids a race
		// condition during deployment where a 404 for a
		// not-fully-propagated asset can get cached by Cloudflare and
		// prevent any users from entire geographic regions from ever
		// being able to load that asset.
		//
		// Assets is backed by in-memory byte arrays, so this is a
		// cheap operation.
		f, err := assets.Assets.Open(r.URL.Path)
		if f != nil {
			defer f.Close()
		}
		if err == nil {
			if isPhabricatorAsset(r.URL.Path) {
				w.Header().Set("Cache-Control", "max-age=300, public")
			} else {
				w.Header().Set("Cache-Control", "immutable, max-age=172800, public")
			}
		}

		fs.ServeHTTP(w, r)
	})))
}
```

For a given request that's for an asset ( `./assets/...`): we tell Cloudflare to cache the asset for two days ([via the `cache-control` header](https://support.cloudflare.com/hc/en-us/articles/115003206852-Understanding-Origin-Cache-Control)s) iff the file path specified via the [URL's `path` component](https://golang.org/pkg/net/url/#URL) is present in our in-memory file system. 

(**Important note**: We only consider the `path` component of the URL - nothing else. This will be important later on).  

## Chunk caching 

For chunks, the caching story is simple. Each chunk is given a unique name based off of its contents (ignoring exceedingly unlikely hash collisions). Since we can assume that each file is uniquely addressable, we can also assume that Cloudflare's cache for that particular asset will only contain the file contents that we expect. 

## app.bundle.js caching

However, `app.bundle.js` runs into issues. The *filename* for the bundle is static across every build of `sourcegraph/sourcegraph`. We have some logic in our html templates that generate unique *requests* for the `app.bundle.js` asset across each build (see [app.html](https://github.com/sourcegraph/sourcegraph/blob/71cb97d88538551e961397a93d74321a8d103d0f/cmd/frontend/internal/app/ui/app.html#L60) and [tmpl.go](https://github.com/sourcegraph/sourcegraph/blob/71cb97d88538551e961397a93d74321a8d103d0f/cmd/frontend/internal/app/ui/tmpl.go#L35-L64) for that logic) by using query paramters.

Example: 

- version 1 of `sourcegraph/sourcegraph` generates requests like `./assets/app.bundle.js?v1`
- version 2 of `sourcegraph/sourcegraph` generates requests like `./assets/app.bundle.js?v2`
- etc.

These query parameters are ignored by the `http.Handler` above (which only considers the URL's `path` → `./assets/app.bundle.js` ). 

So, ignoring a cache sitting in front of the instance, the same instance (regardless of its version) will map both `./assets/app.bundle.js?v1` and
`./assets/app.bundle.js?vFOO` to the same underlying asset: `app.bundle.js`. 

Despite this, Cloudflare's caching would still only contain the correct contents for a given "version" of `app.bundle.js` if we could guarantee that that requests for a specific version were **routed to the correct version** of the `sourcegraph/sourcegraph` instance (e.g. `./assets/app.bundle.js?v2` **is only ever routed to** `sourcegraph/sourcegraph@v2)`.  

### Where it all goes wrong 

This is not the case during one scenario: deployments. `sourcegraph-frontend` uses `RollingDeployments` ([link](https://github.com/sourcegraph/deploy-sourcegraph/blob/76bef1d0dcfd9228eb43811d4ef4446cc88452a1/base/frontend/sourcegraph-frontend.Deployment.yaml#L18-L22)), which means that both the "old" and "new" versions of `sourcegraph-frontend` run alongside each other until the "new" version of the `sourcegraph-frontend` pod is healthy.  

During this time, requests can be routed to **either** the old or new versions of the `sourcegraph-frontend` pods. This allows for the following scenario:

1. A new deployment of `sourcegraph-frontend` starts rolling out, let's call it `v2`.
2. One of the `frontend` pods is running `v2`, all the others are still running `v1` as the rollout hasn't fully propagated yet
3. A user loads a page and gets the latest HTML (from `v2`), his browser asks for `app.bundle.js?v2`
4. The Cloudflare cache does not have `app.bundle.js?v2` yet, so it requests from the origin. BUT the deployment hasn't rolled out fully yet, so the request makes its way to a server running `v1`. Now `app.bundle.js` from a v1 server is cached by cloudflare as `app.bundle.js?v2`. Due to our caching logic in the handler,  Cloudflare is going to serve the `v1` version of the `app.bundle.js` asset for `app.bundle.js?v2` requests for two days. 😢
5. The deployment rollout completes. Only the `v2` frontend pods are running now. 
5. The `v1` version of the `app.bundle.js` now makes requests for `v1` versions of chunk files that aren't in the Cloudflare cache. Sourcegraph 404's when Cloudflare makes requests for those chunkfiles, resulting in a broken user experience. The app.bundle.js file will remain in the cloudflare cache for two days (sourcegraph returned an asset file for it), but all the requests that it makes for chunkfiles will fail. 

**There is no resolution that the user can perform for this**. Clearing your browser cache/incognito will do nothing since this is what Cloudflare is serving for us on our behalf. 

Sorry for the long-winded explanation, but this is full description for what's going on with the issues documented in [https://sourcegraph.slack.com/archives/C01RNLHGU3Z](https://sourcegraph.slack.com/archives/C01RNLHGU3Z) 

# Resolution

We should give `app.bundle.js` a filename that's unique based of a hash of its contents like we do for chunkfiles. This way, outdated references to old versions of our app bundle will simply 404 (because file name would change with each new version of sourcegraph). 

My rough sketch for how this could work:

1. Webpack writes the current name of the `app.bundle-[hash].js` file to some known file like `/bundle-name.json` whenever the file contents change 
2. We can lift this into the Go template logic that [generates app.html](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/frontend/internal/app/ui/app.html#L60): by something like :

```go

type Bundle struct {
        Name string
}

//go:embed bundle-name.json
var bundleName string

tmpl := template.New("...")
tmpl := tmpl.MustParse("...")

b := Bundle{bundleName}

err := tmpl.Execute(os.Stdout, b)
if err != nil {
	...
}

```

3. Changing the javascript references in [https://sourcegraph.com/search?q=repo:^github\.com/sourcegraph/sourcegraph%24+app.bundle.js&patternType=literal](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+app.bundle.js&patternType=literal) ([https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/web/src/integration/context.ts?subtree=true](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/web/src/integration/context.ts?subtree=true) and [https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/web/package.json?subtree=true](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/web/package.json?subtree=true)) to get the bundle file path from `bundle-name.json`

# Conclusion

Sorry for the long-winded explanation, but this is full explanation for what's going on with the issues documented in [https://sourcegraph.slack.com/archives/C01RNLHGU3Z](https://sourcegraph.slack.com/archives/C01RNLHGU3Z) 

Thanks so much to @slimsag and @pecigonzalo for providing valuable insight here.

cc @sourcegraph/frontend-platform @sourcegraph/core-application Do the suggested actions in the conclusion make sense? If so, I can probably tackle this as long as someone can point me to how I'd get webpack to generate a file that contains the current hashed-filename for the app.bundle.js file.

**Comment by felixfbecker** [comment]
Wow, that's a grade A+ investigation @ggilmore 💯 🤯

**Comment by umpox** [comment]
@ggilmore Thank you for taking the time to document this problem 🙏 

For generating a list of our hashed files we should be able to use https://github.com/shellscape/webpack-manifest-plugin. I'm just investigating if there are any cleaner ways for us to achieve this

**Comment by pecigonzalo** [comment]
Amazing investigation @ggilmore. Really good find on the rolling deployments! that would also make it more confusing to find and potentially not happen as often with the old deployment schedule.

**Comment by umpox** [comment]
I made a draft PR adding `webpack-manifest-plugin`. Would be interested to get some 👀  This seems like the best approach with our current setup, I think it just needs some template logic to inject the correct bundle filename

https://github.com/sourcegraph/sourcegraph/pull/19479

**Comment by unknwon** [comment]
That problem would be really hard to find out! Nice write up ❤️ and the **Resolution** makes total sense to me.

**Comment by umpox** [comment]
Current state of play:

Backend PR to add initial backend logic to fetch and route to the updated bundle: https://github.com/sourcegraph/sourcegraph/pull/19509

Frontend PR to enable hashing and output a manifest file to be consumed by the backend: https://github.com/sourcegraph/sourcegraph/pull/19648

These PRs have since been merged together. 
Outstanding work:
- Update backend logic to consume the new generated manifest file
- Finally, test everything works as expected when running the application with a production bundle

**Comment by umpox** [comment]
Updated state of play:
- Backend logic has been updated to fix some issues with loading the webpack manifest in local dev. Also fixes performance concerns for production. Thanks @eseliger!
- Currently there seems to be a failing backend test that we still need to solve

From a frontend platform team perspective, there's not much more frontend work we can do to support this and we'll be actively involved in the redesign from next week so we will have less time capacity.

Can @sourcegraph/distribution or @sourcegraph/core-application take ownership of the remaining work and/or review the current PR: https://github.com/sourcegraph/sourcegraph/pull/19479

**Comment by umpox** [comment]
Update:

From a frontend-platform perspective, PR is ready to be released https://github.com/sourcegraph/sourcegraph/pull/19479

Can @sourcegraph/core-application release this?

**Comment by pdubroy** [comment]
Hey everyone! I was talking to @RafLeszczynski about this issue and we thought it would be good idea to do a postmortem. This was part of an [ops incident](https://docs.google.com/document/d/1dtrOHs5STJYKvyjigL1kMm6u-W0mlyRSyVxPfKIOfEw/edit#heading=h.a9hwyy2iulnf), and it's a good time for us to revisit our processes and find ways to improve.

I've started a doc here: https://docs.google.com/document/d/10JnzVhwuRhlF8ViphPm4XXAew-sflUhnKjKEqgeG28M/edit#.  @ggilmore @umpox @unknwon @eseliger — can I get your help in filling in the details?

**Comment by unknwon** [comment]
@pdubroy - Hey, the postmortem comes as a surprise to me. I feel I don't have the context of "having a postmortem" here?

**Comment by pdubroy** [comment]
@unknwon - in general the idea is to use ops incidents as an opportunity to improve our processes, tools, etc — see [Postmortem Culture: Learning from Failure](https://sre.google/sre-book/postmortem-culture/). Typically teams will have a set of criteria for which incidents should have a postmortem. In this case, the resolution time was longer than we'd target for a P1 issue, so it's a good opportunity to look for improvements we can make to our process.

Let me emphasize here that in a healthy post-mortem culture, we assume that everyone involved had good intentions and did the best they could with the information/tools they had. 🙂
